docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,888 @@
+
+  - *Complete Mode* - The entire updated Result Table will be written to the external storage.
+
+  - *Append Mode* - Only the new rows appended in the Result Table since the last trigger will be written to the external storage. This is applicable only on queries where existing rows in the Result Table is not expected to change.
+  
+  - *Update Mode* - Only the rows that were updated in the Result Table since the last trigger wil be written to the external storage (not available yet in Spark 2.0). Note that this is different from the Complete Mode in that this mode does not output the rows that are not changed.
+
+Note that each mode is applicable on certain types of queries. This is discussed in detail later.
+
+To illustrate the use of this model, let’s understand the model in context of 
+the Quick Example above. The first `lines` DataFrame is the input table, and 
+the final `wordCounts` DataFrame is the result table. Note that the query on 
+streaming `lines` DataFrame to generate `wordCounts` is *exactly the same* as 
+it would be a static DataFrame. However, when this query is started, Spark 
+will continuously check for new data from the socket connection. If there is 
+new data, Spark will run an “incremental” query that combines the previous 
+running counts with the new data to compute updated counts, as shown below.
+
+![Model](img/structured-streaming-example-model.png)
+
+This model is significantly different from many other stream processing 
+engines. Many streaming system require the user to maintain running 
+aggregations themselves, thus having the reason about fault-tolerance, and 
+data consistency (at-least-once, or at-most-once, or exactly-once). In this 
+model, Spark is responsible for updating the Result Table when there is new 
+data, thus relieving the users from reasoning about it. As an example, let’s 
+see how this model handles event-time based processing and late arriving data.
+
+## Handling Event-time and Late Data
+Event-time is the time embedded in the data itself. For many applications, you may want to do operate using this event-time. For example, if you want to get the number of events generated by IoT devices every minute, then you probably want to use the time when the data was generated (that is, event-time in the data), rather than the time Spark receives them. This event-time is very naturally expressed in this model -- each event from the devices is a row in the table, and event-time a column value in the row. This allows window-based aggregations (e.g. number of event every minute) to be just a special type of grouping and aggregation on the even-time column -- each time window is a group and each row can belong to multiple windows/groups. Therefore, such event-time-window-based aggregation queries can be defined consistently on both a static dataset (e.g. from collected device events logs) as well as on streaming dataset, making the life of the user much easier.
```

Comment:
want to do operate => want to operate
on streaming dataset => on a streaming dataset